### PR TITLE
Add ECAPA2 to VoxCeleb

### PR DESCRIPTION
**Pull request by othman-istaiteh**
## What does this PR do?

This PR implements the ECAPA2 model architecture and its corresponding training recipe for VoxCeleb.

**Key Additions:**
* **`speechbrain/lobes/models/ECAPA2.py`:** Implementation of the ECAPA2 architecture and `SubCenterClassifier`.
* **`speechbrain/nnet/losses.py`:** Added `JeffreysLoss` for embedding regularization. 
* **VoxCeleb Recipe (`recipes/VoxCeleb/SpeakerRec/`)**: 
  * Added `train_ecapa2.yaml` and `verification_ecapa2.yaml`.
  * Updated `train_speaker_embeddings.py` and `speaker_verification_cosine.py` to support the new model and pipeline requirements. 
  * Handled backward compatibility natively; existing models (X-Vector, ResNet, ECAPA-TDNN) run without modification.

**Testing & Validation:**
* Added ECAPA2 testing vectors to `tests/recipes/VoxCeleb.csv`.
* Ran `pytest tests` to ensure existing functionality remains intact.
* Passed all doctests.
* Ran `pre-commit run -a` to verify strict code formatting and linting.

**Performance:**

*Trained on VoxCeleb 1 + VoxCeleb 2:*
* **VoxCeleb1-O:** 0.60% EER (with s-norm) / 0.70% EER (without s-norm)

*Trained on VoxCeleb 2 only (tested without s-norm):*
* **VoxCeleb1-O:** 0.79% EER
* **VoxCeleb1-E:** 1.00% EER
* **VoxCeleb1-H:** 1.76% EER

Fixes N/A

**Breaking changes:** None. Backward compatibility is maintained for existing VoxCeleb scripts.

<details open>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+out_n_neurons: 5994 # Change to 7205 if training on Vox1 + Vox2
+
+num_workers: 4
+dataloader_options:
+    batch_size: !ref <batch_size>
+    shuffle: !ref <shuffle>
+    num_workers: !ref <num_workers>
+
+# Functions
+use_tacotron2_mel_spec: True
```
Comment (TParcollet) [inline recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml:82]: Why using this? Is there a reason for not using standard Mels?

File: recipes/VoxCeleb/SpeakerRec/hparams/verification_ecapa2.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# ################################
+# Model: Speaker Verification Baseline for ECAPA2
+# Acknowledgment: The source code is derived from the Kiwano toolkit.
```
Comment (TParcollet) [inline recipes/VoxCeleb/SpeakerRec/hparams/verification_ecapa2.yaml:3]: Add author name for tracking.

File: recipes/VoxCeleb/SpeakerRec/README.md [hunk 3]
```diff
@@ -87,13 +96,16 @@ Below results are all obtained with the official verification split of voxceleb1
 |-----------------|------------|------| -----|
 | Xvector + PLDA  | VoxCeleb 1,2 | 3.23% | https://www.dropbox.com/sh/ab1ma1lnmskedo8/AADsmgOLPdEjSF6wV3KyhNG1a?dl=0 |
 | ECAPA-TDNN      | VoxCeleb 1,2 | 0.80% | https://www.dropbox.com/sh/ab1ma1lnmskedo8/AADsmgOLPdEjSF6wV3KyhNG1a?dl=0 |
+| ECAPA2          | VoxCeleb 1,2 | 0.60% | https://drive.google.com/drive/folders/1cpU5qpCVM30Ip8I85EPM33lsUYPa6S7q?usp=sharing |
```
Comment (TParcollet) [inline recipes/VoxCeleb/SpeakerRec/README.md:99]: Please @Adel-Moumen can we upload this to dropbox?

File: recipes/VoxCeleb/SpeakerRec/speaker_verification_cosine.py [hunk 4]
```diff
@@ -46,7 +46,15 @@ def compute_embedding(wavs, wav_lens):
     embeddings : torch.Tensor
     """
     with torch.no_grad():
-        feats = params["compute_features"](wavs)
+        if (
+            "use_tacotron2_mel_spec" in params
```
Comment (TParcollet) [inline recipes/VoxCeleb/SpeakerRec/speaker_verification_cosine.py:50]: Yes, this is a bit confusing. See above question, i'd prefer if we could use standard Mel. Is there a real difference?

File: speechbrain/lobes/models/ECAPA2.py [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from speechbrain.dataio.dataio import length_to_mask
+
+
+class SubCenterClassifier(nn.Module):
+    """Sub-Center ArcFace Classifier.
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:17]: Docstring isn't explicit enough. I don't know what this is.

File: speechbrain/lobes/models/ECAPA2.py [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+        one_hot.scatter_(1, label.view(-1, 1).long(), 1)
+
+        output = one_hot * phi + (1.0 - one_hot) * cosine_sim
+        output *= self.scale
+
+        return output.unsqueeze(1)
+
+
+class ECAPA2DownsampleBlock(nn.Module):
+    """Downsampling block for ECAPA2."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:99]: Incomplete docstring (args)

File: speechbrain/lobes/models/ECAPA2.py [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+        )
+        self.bn = nn.BatchNorm2d(out_channels)
+
+    def forward(self, x):
+        """Processes the input tensor."""
+        return self.bn(self.conv(x))
+
+
+class ECAPA2SEBlock2d(nn.Module):
+    """2D Squeeze-and-Excitation block."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:118]: Incomplete docstring (args)

File: speechbrain/lobes/models/ECAPA2.py [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+        w = self.relu(w)
+        w = self.bn1(w)
+        w = self.fc2(w)
+        w = self.sigmoid(w)
+        w = w.view([b, c, 1, 1])
+        return x * w
+
+
+class ECAPA2ConvBlock(nn.Module):
+    """Convolutional block with SE processing for ECAPA2 frontend."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:143]: same

File: speechbrain/lobes/models/ECAPA2.py [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+        y = self.relu(y)
+        y = self.bn1(y)
+        y = self.fc2(y)
+        y = self.sigmoid(y)
+        y = y.unsqueeze(2)
+        return x * y
+
+
+class ECAPA2Res2NetConv1d(nn.Module):
+    """Res2Net convolutional block for 1D features."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:207]: same

File: speechbrain/lobes/models/ECAPA2.py [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+            sp = self.batch_norms[i - 1](sp)
+            outputs.append(sp)
+        x = torch.cat(outputs, dim=1)
+        x = self.bn2(self.relu2(self.conv2(x)))
+        x = self.se(x)
+        return x
+
+
+class ECAPA2TDNNBlock(nn.Module):
+    """TDNN block for ECAPA2."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:265]: same

File: speechbrain/lobes/models/ECAPA2.py [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+        self.relu = nn.ReLU()
+        self.bn = nn.BatchNorm1d(out_channels)
+
+    def forward(self, x):
+        """Processes the input tensor."""
+        return self.bn(self.relu(self.conv(x)))
+
+
+class ECAPA2DenseBlock(nn.Module):
+    """Dense convolutional block for ECAPA2."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:285]: same

File: speechbrain/lobes/models/ECAPA2.py [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+        )
+        self.relu = nn.ReLU()
+
+    def forward(self, x):
+        """Processes the input tensor."""
+        return self.relu(self.conv(x))
+
+
+class ECAPA2AttentiveStatPoolingBlock(nn.Module):
+    """Attentive Statistics Pooling for ECAPA2."""
```
Comment (TParcollet) [inline speechbrain/lobes/models/ECAPA2.py:300]: same

File: speechbrain/nnet/losses.py [hunk 13]
```diff
@@ -1988,3 +1988,137 @@ def forward(self, predictions, length=None, reduction=None):
         else:
             loss = laplacian.masked_select(mask).var()
         return -loss
+
+
+class JeffreysLoss(nn.Module):
+    """Computes the Jeffreys Loss, a combination of Cross Entropy, Label Smoothing,
```
Comment (TParcollet) [inline speechbrain/nnet/losses.py:1994]: Can we get a unit test for this new loss please?

**Review by TParcollet — CHANGES_REQUESTED** [review]
Hey! Thank you very much for this recipe! I won't be able to try it because we do not have the voxceleb data. Before finding someone to try it, could you please address the comments?

File: recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml (hunk 1 above)
Comment (othman-istaiteh) [reply recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml:82]: No, I just followed the same used in this recipe https://github.com/speechbrain/speechbrain/blob/develop/recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa_tdnn_mel_spec.yaml

But I can change it to standard mels

Comment (TParcollet) [reply recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml:82]: Oh, this is new, interesting, then ok. Could you try with the standard ones just to see the end result maybe?

Comment (othman-istaiteh) [reply recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml:82]: I am sorry for the late reply
I will do the experiments and get back to you.

Comment (othman-istaiteh) [reply recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml:82]: I double-checked several times using the compute features as ecappa 1, and the results degraded drastically 
compute_features: !new:speechbrain.lobes.features.Fbank
    n_mels: !ref <n_mels>
    left_frames: !ref <left_frames>
    right_frames: !ref <right_frames>
    deltas: !ref <deltas>

is it okay with you to keep it like this ?

Comment (othman-istaiteh) [reply recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa2.yaml:82]: @TParcollet could you please confirm if it's okay to use the Tacotron 2 mel spec?
